### PR TITLE
[cudaaaligner] Improved task scheduling for Banded Myers

### PR DESCRIPTION
**Pull request by ahehn-nv**
* Added NVTX markers to cudaaligner
* Improved scheduling of unbalanced workloads. Larger alignments will be processed first now.
* Minor improvement of `reverse_complement()`

### CODE REVIEW

File: common/base/include/claraparabricks/genomeworks/utils/genomeutils.hpp [hunk 1]
```diff
@@ -143,16 +143,11 @@ inline std::vector<std::string> generate_random_sequences(std::string const& bac
 
 inline void reverse_complement(const char* src, const int32_t length, char* dest)
 {
+    constexpr char lookup[] = {'T', 'G', 'A', 'C'};
```
Comment (mimaric) [inline common/base/include/claraparabricks/genomeworks/utils/genomeutils.hpp:146]: Got it, but a comment would be useful

File: cudaaligner/src/aligner_global_myers_banded.cpp [hunk 2]
```diff
@@ -278,12 +290,20 @@ StatusType AlignerGlobalMyersBanded::align_all()
     {
         result_lengths_d.clear_and_resize(n_alignments);
     }
+    if (get_size(scheduling_index_d) < n_alignments)
+    {
+        scheduling_index_d.clear_and_resize(n_alignments);
+    }
+    scheduling_index_h.resize(n_alignments);
+    std::iota(begin(scheduling_index_h), end(scheduling_index_h), 0);
+    std::sort(begin(scheduling_index_h), end(scheduling_index_h), [&seq_starts_h](int32_t i, int32_t j) { return seq_starts_h[2 * i + 2] - seq_starts_h[2 * i] > seq_starts_h[2 * j + 2] - seq_starts_h[2 * j]; });
```
Comment (mimaric) [inline cudaaligner/src/aligner_global_myers_banded.cpp:299]: Add a comment

File: cudaaligner/src/myers_gpu.cu [hunk 3]
```diff
@@ -765,6 +770,17 @@ myers_compute_scores_edit_dist_banded(
...
 }
 
+__device__ int32_t get_alignment_task(const int32_t* scheduling_index_d, cuda::atomic<int32_t, cuda::thread_scope_device>* scheduling_atomic_d)
+{
+    int32_t sched_idx = 0;
+    if (threadIdx.x == 0)
+    {
+        sched_idx = scheduling_atomic_d->fetch_add(1, cuda::memory_order_relaxed);
+    }
+    sched_idx = __shfl_sync(0xffff'ffffu, sched_idx, 0);
```
Comment (mimaric) [inline cudaaligner/src/myers_gpu.cu:780]: Add a comment

**Review by mimaric — CHANGES_REQUESTED** [review]
If I got it right it is ok, but please add some comments. There is not a single mention that the alignments get sorted by size for balancing reasons, so anyone looking at the code without the context of this PR would have a really hard time

File: common/base/include/claraparabricks/genomeworks/utils/genomeutils.hpp (hunk 1 above)
Comment (ahehn-nv) [reply common/base/include/claraparabricks/genomeworks/utils/genomeutils.hpp:147]: Done.

File: cudaaligner/src/aligner_global_myers_banded.cpp (hunk 2 above)
Comment (ahehn-nv) [reply cudaaligner/src/aligner_global_myers_banded.cpp:302]: Yes, you're right. This needs a comment. Done.

File: cudaaligner/src/myers_gpu.cu (hunk 3 above)
Comment (ahehn-nv) [reply cudaaligner/src/myers_gpu.cu:786]: Done

**Comment by ahehn-nv** [comment]
rerun tests

**Review by mimaric — APPROVED** [review]

**Comment by ohadmo** [comment]
rerun tests

**Comment by ahehn-nv** [comment]
rerun tests

**Review by mimaric — APPROVED** [review]

**Pull request merged**